### PR TITLE
Allow syncing of home option via WP-CLI

### PR DESCRIPTION
**Pull request by gititon**
Allow syncing of home option via wp-cli. See https://github.com/Automattic/jetpack/issues/12924 

Testing: 
This can't be tested easily using the Jetpack Docker setup because it always sets the `host` option to `localhost`, which puts Jetpack in development mode and [disallows sync](https://github.com/Automattic/jetpack/blob/master/packages/sync/src/Actions.php#L117).

Deploy this to Jetpack in a different environment, monitor the sync data received on your wpcom sandbox, and run

`wp option update home https://some-new-domain.com`

Ensure that the value is received accordingly on wpcom.

Learnings:
1) A host without a period in it (e.g., localhost) will cause Jetpack to be in development mode
2) Disable invocation of sync via periodic cron invoked by requests by adding `define('DISABLE_WP_CRON', 'true');` to wp-config.php
3) Initiate sync via wp cli via `wp cron event run jetpack_sync_cron --allow-root`, note that specifying --all instead will run both jetpack_sync_cron and jetpack_full_sync_cron, potentially giving you twice the debugging statements you were expecting

### CODE REVIEW

File: packages/sync/src/modules/Callables.php
Comment (tyxla) [inline packages/sync/src/modules/Callables.php]: Why are we removing this one? What's the impact of this change?

File: packages/sync/src/Main.php
Comment (tyxla) [inline packages/sync/src/Main.php]: Interesting, I have 3 questions here: 

* Why do we need to do this? Ideally, we would just make sure to initialize it only once.
* This seems like material for another PR, how is it related with this one?
* How do we test the impact of this change?

File: tests/php/sync/test_class.jetpack-sync-callables.php
Comment (tyxla) [inline tests/php/sync/test_class.jetpack-sync-callables.php]: This seems wrong - it doesn't seem like we maintain array keys for `Callables::ALWAYS_SEND_UPDATES_TO_THESE_OPTIONS`. Should we use `array_search` to locate the right key instead?

File: tests/php/sync/test_class.jetpack-sync-callables.php
Comment (tyxla) [inline tests/php/sync/test_class.jetpack-sync-callables.php]: If we use `array_search` as suggested above, then we could directly update the existing element, rather than removing the previous one and inserting a new onne.

File: packages/sync/src/modules/Callables.php
Comment (tyxla) [inline packages/sync/src/modules/Callables.php]: Any others we should take care for? For example, unless I'm wrong, I think `paused_themes` and `paused_plugins` should correspond to `wp_paused_themes()` and `wp_paused_plugins`. Should we do a full audit of the "always send updates to these options" corresponding callables?

File: packages/sync/src/modules/Callables.php
Comment (tyxla) [inline packages/sync/src/modules/Callables.php]: `$key` here can be a bit confusing, it's not clear what it really represents.

File: packages/sync/src/modules/Callables.php
Comment (tyxla) [inline packages/sync/src/modules/Callables.php]: We seem to be mutating `$key` here, which can create unexpected results when refactoring the code.

**Review by tyxla — COMMENTED** [review]
I've pushed some commits to fix formatting and tests, and also some minor improvements.

Adding some questions and suggestions, going to address some of them myself.

File: packages/sync/src/modules/Callables.php
Comment (tyxla) [reply packages/sync/src/modules/Callables.php]: Resolved in 3a3060b500b5f94deb3a548994341b445f7e3d1c.

File: packages/sync/src/Main.php
Comment (gititon) [reply packages/sync/src/Main.php]: @tyxla When I was working on this, I discovered that init was called on it twice. It made things clearer once it was only invoked once. I'm happy to hold off and submit this in a separate PR if you prefer?

File: packages/sync/src/modules/Callables.php
Comment (gititon) [reply packages/sync/src/modules/Callables.php]: @tyxla , the goal of this PR was to ensure that changing the `home` option via WP-CLI could sync, https://github.com/Automattic/jetpack/issues/12924, so that's what I tested / focused on. I haven't explored other options/callables, and would like to get this PR merged. What about exploring additional options/callables in a subsequent PR or adding them as necessary?

File: packages/sync/src/Main.php
Comment (tyxla) [reply packages/sync/src/Main.php]: > I'm happy to hold off and submit this in a separate PR if you prefer?

That's my preference, but it's not super strong, so I'll defer to you to make the call 😉 I just think it's something that deserves a special attention, dedicated testing, maybe a couple of tests too.

File: packages/sync/src/modules/Callables.php
Comment (tyxla) [reply packages/sync/src/modules/Callables.php]: Another PR sounds good to me! I'd be fine with adding a `@TODO` in the docblock or inside the array so we keep track of what needs to happen here.

File: tests/php/sync/test_class.jetpack-sync-callables.php
Comment (gititon) [reply tests/php/sync/test_class.jetpack-sync-callables.php]: @tyxla , `array_key_exists` is the correct function here, because we want to see if the array/map `Callables::OPTION_NAMES_TO_CALLABLE_NAMES` has a key `$option`. If it does, that means the callable name differs from the option name, and we replace the option name with the callable name in the list of callables to always update.

File: tests/php/sync/test_class.jetpack-sync-callables.php
Comment (gititon) [reply tests/php/sync/test_class.jetpack-sync-callables.php]: Please see https://github.com/Automattic/jetpack/pull/13015/files#r304931686

File: tests/php/sync/test_class.jetpack-sync-callables.php
Comment (tyxla) [reply tests/php/sync/test_class.jetpack-sync-callables.php]: I think we're talking about different things. 

`$always_updated` uses `Callables::OPTION_NAMES_TO_CALLABLE_NAMES`, and the latter is an array with numeric keys, as defined in the class. So basically, how would we have `$option` as a key there? And then we're trying to unset that key?

File: tests/php/sync/test_class.jetpack-sync-callables.php
Comment (tyxla) [reply tests/php/sync/test_class.jetpack-sync-callables.php]: I might be misunderstanding how this works, but see https://github.com/Automattic/jetpack/pull/13015/files#r304934427

**Review by tyxla — APPROVED** [review]
Code looks great, thanks! Note that I haven't tested after the last changes, would be great if I could ask @timmyc or @roccotripaldi to give it a spin!

**Review by roccotripaldi — APPROVED** [review]
This seems good to me. I tested on my wpsandbox.me site

**Comment by nickdaugherty** [comment]
Thank you @gititon!

File: packages/sync/src/modules/Callables.php
Comment (nickdaugherty) [inline packages/sync/src/modules/Callables.php]: @gititon the `is_admin()` function doesn't check the current user, it checks if it's an admin request (i.e., to an url under `/wp-admin`)

File: packages/sync/src/modules/Callables.php
Comment (nickdaugherty) [inline packages/sync/src/modules/Callables.php]: This check is still preventing callables from being synced on shutdown for WP CLI commands - maybe we just need a check for `defined( 'WP_CLI' ) && WP_CLI`?

**Comment by gititon** [comment]
Hi @nickdaugherty, I just tested it again on master and confirmed that it works.

When I run `wp option update home https://some-new-domain-my-site.com`, I see the new domain name sync over to my wpcom sandbox. 

```
[19-Jul-2019 20:17:05 UTC] [df0f-22278] jetpack_sync_callable
[19-Jul-2019 20:17:05 UTC] Trigger: jetpack_sync_callable
[19-Jul-2019 20:17:05 UTC] Sync args:
Array
(
    [0] => home_url
    [1] => https://some-new-domain-my-site.com
)
```

Is there anything unique about your setup I can try to do to replicate your environment?

**Pull request merged**
